### ISSUE TITLE
Ember build failures cause timeouts

### ISSUE THREAD

**Issue by seanpdoyle** [thoughtbot/ember-cli-rails#240]
Ember builds that fail due to missing or out of date dependencies don't cause Rails errors.

Instead, the request will timeout, in which case all bets are off and the failures that *are* rendered are largely red-herrings.

**Comment by seanpdoyle** [comment]
```
I’ve modified https://github.com/quaertym/ember-cli-dependency-checker/blob/master/index.js to run in `preBuild` (I didn't think errors in its `init` hook would be handled by `buildError`)

I’m forcing a build failure in a project’s `preBuild` by omitting a dependency.

No matter what I do, I can’t get `ember-cli-rails-addon`’s `buildError` to handle the error

GitHub
quaertym/ember-cli-dependency-checker
ember-cli-dependency-checker - Ember CLI addon for checking missing node and bower dependencies before running ember commands
(edited)

rwjblue [10:06 AM] 
why would dep checker want to run in prebuild?

seanpdoyle [10:06 AM] 
to check dependencies before a build?

rwjblue [10:06 AM] 
`init` is before building :wink:

seanpdoyle [10:06 AM] 
Also I was unsure if it was throwing exceptions too early for `ember-cli-rails-addon` to catch them

seanpdoyle [10:07 AM]
I’m also fish out of water and desperately trying anything

rwjblue [10:07 AM] 
but it is also before any other non-build related commands

rwjblue [10:07 AM]
if we move this check to preBuild, we loose much of the power of the dependency checker

rwjblue [10:07 AM]
because then it only checks for things that run builds

seanpdoyle [10:07 AM] 
ok, I buy that

rwjblue [10:07 AM] 
(which is not everything)

rwjblue [10:07 AM]
the key (like with bundler) is that is is ALWAYS ran

seanpdoyle [10:07 AM] 
what can the rails addon do to notice that exception?

seanpdoyle [10:08 AM]
it’s killing the process before it even knows about the rails addon

rwjblue [10:08 AM] 
how is it spawning?

rwjblue [10:08 AM]
then you need to check exit code of the spawned process

rwjblue [10:08 AM]
(and its stderr contains the error output)

seanpdoyle [10:09 AM] 
`ember-cli-rails` currently calls `Kernel.spawn` on

https://github.com/thoughtbot/ember-cli-rails/blob/master/lib/ember-cli/app.rb#L285

which splits `ember build —watch` into its own process

GitHub
thoughtbot/ember-cli-rails
ember-cli-rails - Easily integrate Ember CLI and Rails

seanpdoyle [10:10 AM]
the problem is that the rest of `ember-cli-rails` depends on the rails addon to communicate ember-side errors by catching them in `buildError` and writing them to a file

seanpdoyle [10:11 AM]
TL;DR the gem doesn’t do anything to check if the `spawn` call of `ember build` fails

rondale_sc [10:11 AM] 
@seanpdoyle:  are you able to `node debug $(which ember) build` and catch a `debugger` inside the error hook? (edited)

rwjblue [10:11 AM] 
@seanpdoyle: then you need to spawn a thing that does what you want

rwjblue [10:11 AM]
@seanpdoyle: for example, spawn another ruby script that runs `ember build` and emits the error.json file you need

seanpdoyle [10:11 AM] 
does `ember build` only write bad stuff (errors and whatnot) to `stderr`?

rwjblue [10:11 AM] 
¯\_(ツ)_/¯

rwjblue [10:12 AM]
you tell me

seanpdoyle [10:12 AM] 
What is an `error.json`?

rwjblue [10:12 AM] 
I dunno!

rwjblue [10:12 AM]
https://embercommunity.slack.com/archives/ember-cli/p1446649834002151
Sean Doyleseanpdoyle
the problem is that the rest of `ember-cli-rails` depends on the rails addon to communicate ember-side errors by catching them in `buildError` and writing them to a file
Today at 10:10 AM

seanpdoyle [10:12 AM] 
ok, just making sure its not a std ember-cli thing

rwjblue [10:12 AM] 
I just picked `errors.json` as "to a file"'s name

seanpdoyle [10:12 AM] 
gotcha

seanpdoyle [10:13 AM]
ok, just to clarify, ember-cli-dependency-checker ​*DOES*​ fail too early for the addon to notice. therefore, we need to come up with another way to handle pre-build failures gem-side

rondale_sc [10:13 AM] 
@rwjblue: why do you need to spawn?

seanpdoyle [10:13 AM] 
@rondale_sc: that’s the way the gem currently does it

rwjblue [10:13 AM] 
@rondale_sc: I DON'T!!!
:joy:1  

rwjblue [10:13 AM]
@rondale_sc: AFAIK that is what it was already doing

seanpdoyle [10:13 AM] 
these are all my/our problems, not his

rondale_sc [10:13 AM] 
Oh I see you mean the gem is calling `spawn` to fire off the build, yes?

rwjblue [10:14 AM] 
RE: https://embercommunity.slack.com/archives/ember-cli/p1446649989002167
Sean Doyleseanpdoyle
ok, just to clarify, ember-cli-dependency-checker ​*DOES*​ fail too early for the addon to notice. therefore, we need to come up with another way to handle pre-build failures gem-side
Today at 10:13 AM

rwjblue [10:14 AM]
not quite right

rondale_sc [10:14 AM] 
^ kk, I think I'm following along now.

rwjblue [10:14 AM] 
it isn't about early

rwjblue [10:14 AM]
its that this isn't a build error

rwjblue [10:14 AM]10:14
and only buildErrors are caught by the `buildError` hook
```

**Comment by seanpdoyle** [comment]
Closed by https://github.com/thoughtbot/ember-cli-rails/commit/3703693e22b06df5be31aa9ca7c77c7df8ce8835 and https://github.com/ember-cli/ember-cli/commit/8c93854ae8b9222aff5406c7211d935ceb49c6d6.

Newer versions of `ember-cli` will redirect `stderr` to our build errors file, which `ember-cli-rails` will detect and present to the user.
